File: backend/models/component_logic.py
"""Pydantic models for component logic requests and responses."""

from typing import Dict, Any, List, Optional
from pydantic import BaseModel, EmailStr, validator
import re

# AI Research Models

class UserInfoRequest(BaseModel):
    """Request model for user information validation."""
    full_name: str
    email: str
    company: str
    role: str
    
    @validator('full_name')
    def validate_full_name(cls, v):
        if not v or len(v.strip()) < 2:
            raise ValueError('Full name must be at least 2 characters long')
        return v.strip()
    
    @validator('email')
    def validate_email(cls, v):
        # Basic email validation
        email_pattern = re.compile(r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$')
        if not email_pattern.match(v):
            raise ValueError('Invalid email format')
        return v.lower()
    
    @validator('company')
    def validate_company(cls, v):
        if not v or len(v.strip()) < 1:
            raise ValueError('Company name is required')
        return v.strip()
    
    @validator('role')
    def validate_role(cls, v):
        valid_roles = ["Content Creator", "Marketing Manager", "Business Owner", "Other"]
        if v not in valid_roles:
            raise ValueError(f'Role must be one of: {", ".join(valid_roles)}')
        return v

class ResearchPreferencesRequest(BaseModel):
    """Request model for research preferences configuration."""
    research_depth: str
    content_types: List[str]
    auto_research: bool
    factual_content: bool = True  # Default to True
    
    @validator('research_depth')
    def validate_research_depth(cls, v):
        valid_depths = ["Basic", "Standard", "Deep", "Comprehensive"]
        if v not in valid_depths:
            raise ValueError(f'Research depth must be one of: {", ".join(valid_depths)}')
        return v
    
    @validator('content_types')
    def validate_content_types(cls, v):
        valid_types = ["Blog Posts", "Social Media", "Technical Articles", "News", "Academic Papers"]
        if not v:
            raise ValueError('At least one content type must be selected')
        for content_type in v:
            if content_type not in valid_types:
                raise ValueError(f'Invalid content type: {content_type}')
        return v

class ResearchRequest(BaseModel):
    """Request model for research processing."""
    topic: str
    preferences: ResearchPreferencesRequest
    
    @validator('topic')
    def validate_topic(cls, v):
        if not v or len(v.strip()) < 3:
            raise ValueError('Topic must be at least 3 characters long')
        return v.strip()

class UserInfoResponse(BaseModel):
    """Response model for user information validation."""
    valid: bool
    user_info: Optional[Dict[str, Any]] = None
    errors: List[str] = []

class ResearchPreferencesResponse(BaseModel):
    """Response model for research preferences configuration."""
    valid: bool
    preferences: Optional[Dict[str, Any]] = None
    errors: List[str] = []

class ResearchResponse(BaseModel):
    """Response model for research processing."""
    success: bool
    topic: str
    results: Optional[Dict[str, Any]] = None
    error: Optional[str] = None

# Personalization Models

class ContentStyleRequest(BaseModel):
    """Request model for content style configuration."""
    writing_style: str
    tone: str
    content_length: str
    
    @validator('writing_style')
    def validate_writing_style(cls, v):
        valid_styles = ["Professional", "Casual", "Technical", "Conversational", "Academic"]
        if v not in valid_styles:
            raise ValueError(f'Writing style must be one of: {", ".join(valid_styles)}')
        return v
    
    @validator('tone')
    def validate_tone(cls, v):
        valid_tones = ["Formal", "Semi-Formal", "Neutral", "Friendly", "Humorous"]
        if v not in valid_tones:
            raise ValueError(f'Tone must be one of: {", ".join(valid_tones)}')
        return v
    
    @validator('content_length')
    def validate_content_length(cls, v):
        valid_lengths = ["Concise", "Standard", "Detailed", "Comprehensive"]
        if v not in valid_lengths:
            raise ValueError(f'Content length must be one of: {", ".join(valid_lengths)}')
        return v

class BrandVoiceRequest(BaseModel):
    """Request model for brand voice configuration."""
    personality_traits: List[str]
    voice_description: Optional[str] = None
    keywords: Optional[str] = None
    
    @validator('personality_traits')
    def validate_personality_traits(cls, v):
        valid_traits = ["Professional", "Innovative", "Friendly", "Trustworthy", "Creative", "Expert"]
        if not v:
            raise ValueError('At least one personality trait must be selected')
        for trait in v:
            if trait not in valid_traits:
                raise ValueError(f'Invalid personality trait: {trait}')
        return v
    
    @validator('voice_description')
    def validate_voice_description(cls, v):
        if v and len(v.strip()) < 10:
            raise ValueError('Voice description must be at least 10 characters long')
        return v.strip() if v else None

class AdvancedSettingsRequest(BaseModel):
    """Request model for advanced content generation settings."""
    seo_optimization: bool
    readability_level: str
    content_structure: List[str]
    
    @validator('readability_level')
    def validate_readability_level(cls, v):
        valid_levels = ["Simple", "Standard", "Advanced", "Expert"]
        if v not in valid_levels:
            raise ValueError(f'Readability level must be one of: {", ".join(valid_levels)}')
        return v
    
    @validator('content_structure')
    def validate_content_structure(cls, v):
        valid_structures = ["Introduction", "Key Points", "Examples", "Conclusion", "Call-to-Action"]
        if not v:
            raise ValueError('At least one content structure element must be selected')
        for structure in v:
            if structure not in valid_structures:
                raise ValueError(f'Invalid content structure: {structure}')
        return v

class PersonalizationSettingsRequest(BaseModel):
    """Request model for complete personalization settings."""
    content_style: ContentStyleRequest
    brand_voice: BrandVoiceRequest
    advanced_settings: AdvancedSettingsRequest

class ContentStyleResponse(BaseModel):
    """Response model for content style validation."""
    valid: bool
    style_config: Optional[Dict[str, Any]] = None
    errors: List[str] = []

class BrandVoiceResponse(BaseModel):
    """Response model for brand voice configuration."""
    valid: bool
    brand_config: Optional[Dict[str, Any]] = None
    errors: List[str] = []

class PersonalizationSettingsResponse(BaseModel):
    """Response model for complete personalization settings."""
    valid: bool
    settings: Optional[Dict[str, Any]] = None
    errors: List[str] = []

# Research Utilities Models

class ResearchTopicRequest(BaseModel):
    """Request model for topic research."""
    topic: str
    api_keys: Dict[str, str]
    
    @validator('topic')
    def validate_topic(cls, v):
        if not v or len(v.strip()) < 3:
            raise ValueError('Topic must be at least 3 characters long')
        return v.strip()

class ResearchResultResponse(BaseModel):
    """Response model for research results."""
    success: bool
    topic: str
    data: Optional[Dict[str, Any]] = None
    error: Optional[str] = None
    metadata: Optional[Dict[str, Any]] = None 

# Style Detection Models
class StyleAnalysisRequest(BaseModel):
    """Request model for style analysis."""
    content: Dict[str, Any]
    analysis_type: str = "comprehensive"  # comprehensive, patterns, guidelines

class StyleAnalysisResponse(BaseModel):
    """Response model for style analysis."""
    success: bool
    analysis: Optional[Dict[str, Any]] = None
    patterns: Optional[Dict[str, Any]] = None
    guidelines: Optional[Dict[str, Any]] = None
    error: Optional[str] = None
    timestamp: str

class WebCrawlRequest(BaseModel):
    """Request model for web crawling."""
    url: Optional[str] = None
    text_sample: Optional[str] = None

class WebCrawlResponse(BaseModel):
    """Response model for web crawling."""
    success: bool
    content: Optional[Dict[str, Any]] = None
    metrics: Optional[Dict[str, Any]] = None
    error: Optional[str] = None
    timestamp: str

class StyleDetectionRequest(BaseModel):
    """Request model for complete style detection workflow."""
    url: Optional[str] = None
    text_sample: Optional[str] = None
    include_patterns: bool = True
    include_guidelines: bool = True

class StyleDetectionResponse(BaseModel):
    """Response model for complete style detection workflow."""
    success: bool
    crawl_result: Optional[Dict[str, Any]] = None
    style_analysis: Optional[Dict[str, Any]] = None
    style_patterns: Optional[Dict[str, Any]] = None
    style_guidelines: Optional[Dict[str, Any]] = None
    error: Optional[str] = None
    warning: Optional[str] = None
    timestamp: str 

File: frontend/src/api/componentLogic.ts
// Component Logic API integration
import { AxiosResponse } from 'axios';
import { apiClient } from './client';

// AI Research Interfaces
export interface UserInfoRequest {
  full_name: string;
  email: string;
  company: string;
  role: string;
}

export interface UserInfoResponse {
  valid: boolean;
  user_info?: any;
  errors: string[];
}

export interface ResearchPreferencesRequest {
  research_depth: string;
  content_types: string[];
  auto_research: boolean;
  factual_content: boolean;
}

export interface ResearchPreferencesResponse {
  valid: boolean;
  preferences?: any;
  errors: string[];
}

export interface ResearchRequest {
  topic: string;
  preferences: ResearchPreferencesRequest;
}

export interface ResearchResponse {
  success: boolean;
  topic: string;
  results?: any;
  error?: string;
}

// Personalization Interfaces
export interface ContentStyleRequest {
  writing_style: string;
  tone: string;
  content_length: string;
}

export interface ContentStyleResponse {
  valid: boolean;
  style_config?: any;
  errors: string[];
}

export interface BrandVoiceRequest {
  personality_traits: string[];
  voice_description?: string;
  keywords?: string;
}

export interface BrandVoiceResponse {
  valid: boolean;
  brand_config?: any;
  errors: string[];
}

export interface AdvancedSettingsRequest {
  seo_optimization: boolean;
  readability_level: string;
  content_structure: string[];
}

export interface PersonalizationSettingsRequest {
  content_style: ContentStyleRequest;
  brand_voice: BrandVoiceRequest;
  advanced_settings: AdvancedSettingsRequest;
}

export interface PersonalizationSettingsResponse {
  valid: boolean;
  settings?: any;
  errors: string[];
}

// Research Utilities Interfaces
export interface ResearchTopicRequest {
  topic: string;
  api_keys: Record<string, string>;
}

export interface ResearchResultResponse {
  success: boolean;
  topic: string;
  data?: any;
  error?: string;
  metadata?: any;
}

// AI Research API Functions
export async function validateUserInfo(request: UserInfoRequest): Promise<UserInfoResponse> {
  const res: AxiosResponse<UserInfoResponse> = await apiClient.post('/api/onboarding/ai-research/validate-user', request);
  return res.data;
}

export async function configureResearchPreferences(request: ResearchPreferencesRequest): Promise<ResearchPreferencesResponse> {
  const res: AxiosResponse<ResearchPreferencesResponse> = await apiClient.post('/api/onboarding/ai-research/configure-preferences', request);
  return res.data;
}

export async function processResearchRequest(request: ResearchRequest): Promise<ResearchResponse> {
  const res: AxiosResponse<ResearchResponse> = await apiClient.post('/api/onboarding/ai-research/process-research', request);
  return res.data;
}

export async function getResearchConfigurationOptions(): Promise<any> {
  const res: AxiosResponse<any> = await apiClient.get('/api/onboarding/ai-research/configuration-options');
  return res.data;
}

export async function getResearchPreferences(): Promise<ResearchPreferencesResponse> {
  const res: AxiosResponse<ResearchPreferencesResponse> = await apiClient.get('/api/onboarding/ai-research/preferences');
  return res.data;
}

// Personalization API Functions
export async function validateContentStyle(request: ContentStyleRequest): Promise<ContentStyleResponse> {
  const res: AxiosResponse<ContentStyleResponse> = await apiClient.post('/api/onboarding/personalization/validate-style', request);
  return res.data;
}

export async function configureBrandVoice(request: BrandVoiceRequest): Promise<BrandVoiceResponse> {
  const res: AxiosResponse<BrandVoiceResponse> = await apiClient.post('/api/onboarding/personalization/configure-brand', request);
  return res.data;
}

export async function processPersonalizationSettings(request: PersonalizationSettingsRequest): Promise<PersonalizationSettingsResponse> {
  const res: AxiosResponse<PersonalizationSettingsResponse> = await apiClient.post('/api/onboarding/personalization/process-settings', request);
  return res.data;
}

export async function getPersonalizationConfigurationOptions(): Promise<any> {
  const res: AxiosResponse<any> = await apiClient.get('/api/onboarding/personalization/configuration-options');
  return res.data;
}

export async function generateContentGuidelines(settings: any): Promise<any> {
  const res: AxiosResponse<any> = await apiClient.post('/api/onboarding/personalization/generate-guidelines', settings);
  return res.data;
}

// Research Utilities API Functions
export async function processResearchTopic(request: ResearchTopicRequest): Promise<ResearchResultResponse> {
  const res: AxiosResponse<ResearchResultResponse> = await apiClient.post('/api/onboarding/research/process-topic', request);
  return res.data;
}

export async function processResearchResults(results: any): Promise<any> {
  const res: AxiosResponse<any> = await apiClient.post('/api/onboarding/research/process-results', results);
  return res.data;
}

export async function validateResearchRequest(topic: string, api_keys: Record<string, string>): Promise<any> {
  const res: AxiosResponse<any> = await apiClient.post('/api/onboarding/research/validate-request', { topic, api_keys });
  return res.data;
}

export async function getResearchProvidersInfo(): Promise<any> {
  const res: AxiosResponse<any> = await apiClient.get('/api/onboarding/research/providers-info');
  return res.data;
}

export async function generateResearchReport(results: any): Promise<any> {
  const res: AxiosResponse<any> = await apiClient.post('/api/onboarding/research/generate-report', results);
  return res.data;
} 